Add layout tests for ResourceMonitor

basuke opened the pull request:
#### dcd18c28da6d7a531587b1646dad132a02b55284
<pre>
Add layout tests for ResourceMonitor
<a href="https://bugs.webkit.org/show_bug.cgi?id=286318">https://bugs.webkit.org/show_bug.cgi?id=286318</a>
<a href="https://rdar.apple.com/141840408">rdar://141840408</a>

Reviewed by Ben Nham.

Add minimum layout tests for ResourceMonitor. This covers following basic cases for iframes:
- Not be eligible for monitoring and uses small amount of load.
- Not be eligible for monitoring and uses big amount of load.
- Be eligible for monitoring and uses big amount of load.
- Be eligible for monitoring and uses small amount of load.

For easy and consistent testing, add functions to set various conditions for monitoring and
connect them to Internals and TestRunner. Also needs to

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/iframe-monitor/eligibility-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/eligibility.html: Added.
* LayoutTests/http/tests/iframe-monitor/iframe-unload-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/iframe-unload.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/generate-byte.py: Added.
* LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--2.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/iframe-not-eligible.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/iframe-not-eligible2.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/dom/Document.h:
(WebCore::Document::shouldSkipResourceMonitorThrottling const):
(WebCore::Document::setShouldSkipResourceMonitorThrottling):
* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::ResourceMonitor):
(WebCore::ResourceMonitor::setEligibility):
(WebCore::ResourceMonitor::checkNetworkUsageExcessIfNecessary):
(WebCore::networkUsageThresholdWithRandomNoise): Deleted.
* Source/WebCore/loader/ResourceMonitor.h:
* Source/WebCore/loader/ResourceMonitorChecker.cpp:
(WebCore::networkUsageThresholdWithRandomNoise):
(WebCore::ResourceMonitorChecker::checkEligibility):
(WebCore::ResourceMonitorChecker::setNetworkUsageThreshold):
* Source/WebCore/loader/ResourceMonitorChecker.h:
* Source/WebCore/loader/ResourceMonitorThrottler.cpp:
(WebCore::ResourceMonitorThrottler::setCountPerDuration):
* Source/WebCore/loader/ResourceMonitorThrottler.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setResourceMonitorNetworkUsageThreshold):
(WebCore::Internals::shouldSkipResourceMonitorThrottling const):
(WebCore::Internals::setShouldSkipResourceMonitorThrottling):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/UIProcess/API/C/WKContext.cpp:
(WKContextSetResourceMonitorURLsForTesting):
* Source/WebKit/UIProcess/API/C/WKContext.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::createCompiledContentRuleList):
(WebKit::WebProcessPool::platformLoadResourceMonitorRuleList):
(WebKit::WebProcessPool::platformCompileResourceMonitorRuleList):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::loadOrUpdateResourceMonitorRuleList):
(WebKit::WebProcessPool::setResourceMonitorURLsForTesting):
(WebKit::WebProcessPool::platformLoadResourceMonitorRuleList):
(WebKit::WebProcessPool::platformCompileResourceMonitorRuleList):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::requestResourceMonitorRuleLists):
(WebKit::WebProcessProxy::setResourceMonitorRuleListsIfRequired):
(WebKit::WebProcessProxy::setResourceMonitorRuleLists):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setResourceMonitorContentRuleListAsync):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setResourceMonitorList):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
(WTR::TestRunner::canModifyResourceMonitorList const):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveAsyncMessageFromInjectedBundle):
(WTR::TestController::setResourceMonitorList):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: <a href="https://commits.webkit.org/289467@main">https://commits.webkit.org/289467@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e0edf43798ea6a91d9cb05f320a2bbec3a696411

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87079 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6589 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41437 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/91938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/37818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89129 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/6865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14657 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/91938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/37818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90081 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/6865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78820 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/91938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/6865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33196 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/36935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/6865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34075 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/93825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/14241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/93825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/14445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74676 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/93825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18527 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/18082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/7158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/14260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/19553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/14005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/17447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/15786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on LayoutTests/http/tests/iframe-monitor/eligibility-expected.txt:
```diff
@@ -0,0 +1,10 @@
+
+PASS successfullyParsed is true
+
+TEST COMPLETE
```
This expectation file seems incomplete?

cdumez commented on LayoutTests/http/tests/iframe-monitor/eligibility.html:
```diff
@@ -0,0 +1,48 @@
+description("Test resource monitor.");
+
+onload = async () => {
+    testRunner.waitUntilDone();
```
use `jsTestIsAsync = true;` right after the `description()` call above, instead.

cdumez commented on LayoutTests/http/tests/iframe-monitor/eligibility.html:
```diff
@@ -0,0 +1,48 @@
+    shouldNotBe(`document.querySelector('iframe[name=frame3]').srcdoc`, '""');
+    shouldBe(`document.querySelector('iframe[name=frame4]').srcdoc`, '""');
+
+    testRunner.notifyDone();
```
Should use `finishJSTest()` when using `js-test.js`

cdumez commented on LayoutTests/http/tests/iframe-monitor/eligibility.html:
```diff
@@ -0,0 +1,48 @@
+
+    const result = await setup();
+    if (!result) {
+        testRunner.notifyDone();
```
Should use `finishJSTest()` when using `js-test.js`

cdumez commented on LayoutTests/http/tests/iframe-monitor/iframe-unload.html:
```diff
@@ -0,0 +1,39 @@
+description("Test iframe with huge resource usage is unloaded.");
+
+onload = async () => {
+    testRunner.waitUntilDone();
```
Same issues with this test.

bnham commented on LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js:
```diff
@@ -0,0 +1,50 @@
+        if (!iframe)
+            reject();
+
+        const ticket = setInterval(async () => {
```
Seems a bit weird to use promises, `setTimeout` and `setInterval` here. Am I missing something? An async function is basically another way of writing a function without explicitly using promises, and setting both repeating (setInterval) and one shot (setTimeout) timers seems redundant.

Do you mean to write something like:

```
async function waitUntilUnload(name) {
    const iframe = document.querySelector(`iframe[name=${name}]`);
    if (!iframe)
      throw new Error("iframe doesn't exist"); // throw in async gets turned in to a promise rejection
    
    while (iframe.srcdoc) {
      await pause(100);
    }
}
```

basuke replied:
Right. The script gradually converted to async/await style so that here the mixture was leftover. I'll make it the change.

bnham commented on LayoutTests/platform/gtk/TestExpectations:
```diff
@@ -1949,3 +1949,5 @@ http/tests/cookies/same-site/popup-cross-site-post.html [ Failure ]
 
 webkit.org/b/213782 webanimations/accelerated-animation-with-easing.html [ Failure Pass ]
+
+webkit.org/b/286318 http/tests/iframe-monitor [ Skip ]
```
To match the existing pattern, it's preferable to mark the test as `[ Skip ]` in the top-level TestExpectations file (which it looks like you're already doing), so you don't need to touch the gtk, win, and wpe TestExpectations files.

Assuming that this works on all wk2 ports, then the only platform-specific expectation change should be a `[ Pass ]` in the wk2 expectations file. If this only works in Cocoa WK2 ports, then you should add `[ Pass ]` to the `mac-wk2` and `ios` expectations files.

basuke replied:
Okay.

bnham commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -64,6 +56,7 @@ void ResourceMonitor::setEligibility(Eligibility eligibility)
         return;
 
     m_eligibility = eligibility;
+    RESOURCEMONITOR_RELEASE_LOG("The frame is%s eligible.", (eligibility != Eligibility::Eligible ? " not" : ""));
```
nit: missing a space before `%s`.

I don't remember the exact rules for `os_log` (which is what this compiles down to). In theory, a string with static storage duration should be emitted when using a bare `%s`. But I'm not sure if this works if the string literal is the result of a ternary expression because the way that work might be via something like `__builtin_constant_p(x)`.

You probably want to explicitly use `%" PUBLIC_LOG_STRING` instead to be sure (which compiles down to `%{public}s`:

```
RESOURCEMONITOR_RELEASE_LOG("The frame is %" PUBLIC_LOG_STRING " eligible.", (eligibility != Eligibility::Eligible ? " not" : ""));
```

bnham commented on Source/WebCore/loader/ResourceMonitorChecker.cpp:
```diff
@@ -91,6 +94,8 @@ ResourceMonitor::Eligibility ResourceMonitorChecker::checkEligibility(const Cont
     ASSERT(m_ruleList);
 
     auto matched = m_ruleList->processContentRuleListsForResourceMonitoring(info.resourceURL, info.mainDocumentURL, info.frameURL, info.type);
+    RESOURCEMONITOR_RELEASE_LOG("The url is%s eligible: %s", (!matched ? " not" : ""), info.resourceURL.string().utf8().data());
```
Same nit here, there's a space missing before the `%s`.

The second %s should be marked with `SENSITIVE_LOG_STRING` to align with our standard data logging privacy practices (read about OS logging in the engineering wiki for details):

```
RESOURCEMONITOR_RELEASE_LOG("The url is %" PUBLIC_LOG_STRING " eligible: %" SENSITIVE_LOG_STRING, (!matched ? " not" : ""), info.resourceURL.string().utf8().data());
```

bnham reviewed (approved):
Approved with some nits

basuke committed:
Add layout tests for ResourceMonitor
https://bugs.webkit.org/show_bug.cgi?id=286318
rdar://141840408

Reviewed by Ben Nham.

Add minimum layout tests for ResourceMonitor. This covers following basic cases for iframes:
- Not be eligible for monitoring and uses small amount of load.
- Not be eligible for monitoring and uses big amount of load.
- Be eligible for monitoring and uses big amount of load.
- Be eligible for monitoring and uses small amount of load.

For easy and consistent testing, add functions to set various conditions for monitoring and
connect them to Internals and TestRunner. Also needs to

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/iframe-monitor/eligibility-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/eligibility.html: Added.
* LayoutTests/http/tests/iframe-monitor/iframe-unload-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/iframe-unload.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/generate-byte.py: Added.
* LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/iframe--eligible--2.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/iframe-not-eligible.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/iframe-not-eligible2.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/dom/Document.h:
(WebCore::Document::shouldSkipResourceMonitorThrottling const):
(WebCore::Document::setShouldSkipResourceMonitorThrottling):
* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::ResourceMonitor):
(WebCore::ResourceMonitor::setEligibility):
(WebCore::ResourceMonitor::checkNetworkUsageExcessIfNecessary):
(WebCore::networkUsageThresholdWithRandomNoise): Deleted.
* Source/WebCore/loader/ResourceMonitor.h:
* Source/WebCore/loader/ResourceMonitorChecker.cpp:
(WebCore::networkUsageThresholdWithRandomNoise):
(WebCore::ResourceMonitorChecker::checkEligibility):
(WebCore::ResourceMonitorChecker::setNetworkUsageThreshold):
* Source/WebCore/loader/ResourceMonitorChecker.h:
* Source/WebCore/loader/ResourceMonitorThrottler.cpp:
(WebCore::ResourceMonitorThrottler::setCountPerDuration):
* Source/WebCore/loader/ResourceMonitorThrottler.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setResourceMonitorNetworkUsageThreshold):
(WebCore::Internals::shouldSkipResourceMonitorThrottling const):
(WebCore::Internals::setShouldSkipResourceMonitorThrottling):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/UIProcess/API/C/WKContext.cpp:
(WKContextSetResourceMonitorURLsForTesting):
* Source/WebKit/UIProcess/API/C/WKContext.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::createCompiledContentRuleList):
(WebKit::WebProcessPool::platformLoadResourceMonitorRuleList):
(WebKit::WebProcessPool::platformCompileResourceMonitorRuleList):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::loadOrUpdateResourceMonitorRuleList):
(WebKit::WebProcessPool::setResourceMonitorURLsForTesting):
(WebKit::WebProcessPool::platformLoadResourceMonitorRuleList):
(WebKit::WebProcessPool::platformCompileResourceMonitorRuleList):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::requestResourceMonitorRuleLists):
(WebKit::WebProcessProxy::setResourceMonitorRuleListsIfRequired):
(WebKit::WebProcessProxy::setResourceMonitorRuleLists):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setResourceMonitorContentRuleListAsync):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setResourceMonitorList):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
(WTR::TestRunner::canModifyResourceMonitorList const):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveAsyncMessageFromInjectedBundle):
(WTR::TestController::setResourceMonitorList):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: https://commits.webkit.org/289467@main
